File: igvc_ui/src/lib/protocol/message_accumulator.ts
import { Endianness, MessageType, MessageWrapper, makeMessageWrapper } from "../types";
import { crc32 } from "./crc32";

const Magic = new Uint8Array([0x49, 0x47, 0x56, 0x43]);

const MagicSize = 4;
const HeaderSize = MagicSize + 6;
const CrcSize = 4;

export class MessageAccumulator {
    private buffer: Uint8Array;
    private length = 0;

    constructor(
        private readonly endianness: Endianness,
        private readonly onMessage: (msg: MessageWrapper) => void,
        initialCapacity = 4096,
        private readonly maxMessageLength = 512 * 1024,
        private readonly maxBufferSize = 1024 * 1024
    ) {
        this.buffer = new Uint8Array(initialCapacity);
    }

    append(bytes: Uint8Array) {
        if (this.length + bytes.length > this.maxBufferSize) {
            throw new Error("MessageAccumulator buffer overflow");
        }

        this.ensureCapacity(this.length + bytes.length);
        this.buffer.set(bytes, this.length);
        this.length += bytes.length;

        this.process();
    }

    private ensureCapacity(size: number) {
        if (this.buffer.length >= size) return;

        const next = new Uint8Array(Math.max(size, this.buffer.length * 2));
        next.set(this.buffer.subarray(0, this.length));
        this.buffer = next;
    }

    private process() {
        let offset = 0;
        const view = new DataView(this.buffer.buffer);

        while (this.length - offset >= HeaderSize) {
            // Check magic
            if (!this.matchesMagic(offset)) {
                offset += 1;
                continue;
            }

            const little = this.endianness === Endianness.Little;
            const type = view.getUint16(offset + MagicSize, little);
            const payloadLength = view.getInt32(offset + MagicSize + 2, little);

            if (payloadLength < 0 || payloadLength >= this.maxMessageLength) {
                throw new Error(`Invalid message length ${payloadLength}`);
            }

            const frameSize = HeaderSize + payloadLength + CrcSize;
            if (this.length - offset < frameSize) break;

            const crcStart = offset + HeaderSize + payloadLength;
            const receivedCrc = view.getUint32(crcStart, true);

            const crcData = this.buffer.subarray(offset, offset + HeaderSize + payloadLength);
            const computedCrc = crc32(crcData);

            if (computedCrc !== receivedCrc) {
                throw new Error("CRC mismatch in message");
            }

            const payload = this.buffer.slice(
                offset + HeaderSize,
                offset + HeaderSize + payloadLength
            );

            this.onMessage(makeMessageWrapper(type as MessageType, payload));

            offset += frameSize;
        }

        // Compact buffer
        if (offset > 0) {
            this.buffer.copyWithin(0, offset, this.length);
            this.length -= offset;
        }
    }

    private matchesMagic(offset: number): boolean {
        for (let i = 0; i < MagicSize; i++) {
            if (this.buffer[offset + i] !== Magic[i]) return false;
        }
        return true;
    }
}

File: igvc_csharp/src/Utils/Messages/MessageAccumulator.cs
using System.Buffers.Binary;
using System.IO.Hashing;
using igvc_csharp.Core;

namespace igvc_csharp.Utils.Messages;

/// <summary>
/// Accumulates raw bytes and extracts framed messages.
///
/// Format:
/// [ ushort type ][ int length ][ payload ][ uint crc32c ]
///  
/// CRC covers the (type + length + payload)
/// </summary>
public sealed class MessageAccumulator(
    Endianness endianness,
    Action<MessageWrapper> onMessage,
    int initialCapacity = 4096,
    int maxMessageLength = 512 * 1024, // 512 KB
    int maxBufferSize = 1024 * 1024 // 1024 KB
)
{
    private readonly MemoryStream _stream = new(initialCapacity);

    private const int MagicSize = 4;
    private const int HeaderSize = MagicSize + 6;
    private const int CrcSize = 4;
    
    public void Append(ReadOnlySpan<byte> bytes)
    {
        if (_stream.Length + bytes.Length > maxBufferSize)
        {
            throw new InvalidOperationException("MessageAccumulator buffer overflow");
        }
        
        _stream.Write(bytes);
        Process();
    }

    private void Process()
    {
        _stream.Position = 0;

        while (_stream.Length - _stream.Position >= HeaderSize)
        {
            var buffer = _stream.GetBuffer();
            var span = buffer.AsSpan((int)_stream.Position);

            if (!span[..MagicSize].SequenceEqual(Configuration.NetworkingMagic))
            {
                // skip until magic
                _stream.Position += 1;
                continue;
            }
            
            var type = endianness == Endianness.Little
                ? BinaryPrimitives.ReadUInt16LittleEndian(span[MagicSize..])
                : BinaryPrimitives.ReadUInt16BigEndian(span[MagicSize..]);
            var length = endianness == Endianness.Little
                ? BinaryPrimitives.ReadInt32LittleEndian(span[(MagicSize + 2)..])
                : BinaryPrimitives.ReadInt32BigEndian(span[(MagicSize + 2)..]);

            if (length < 0 || length >= maxMessageLength)
            {
                throw new InvalidOperationException($"Invalid message length {length}");
            }

            var frameSize = HeaderSize + length + CrcSize;
            if (_stream.Length - _stream.Position < frameSize)
            {
                // Wait for more data;
                break;
            }

            var payloadSpan = span.Slice(HeaderSize, length);
            var receivedCrc = BinaryPrimitives.ReadUInt32LittleEndian(span.Slice(HeaderSize + length, CrcSize));
            Span<byte> crcData = stackalloc byte[HeaderSize + length];
            span[..HeaderSize].CopyTo(crcData);
            payloadSpan.CopyTo(crcData[HeaderSize..]);
            var computedCrc = Crc32.HashToUInt32(crcData);
            if (computedCrc != receivedCrc)
            {
                throw new InvalidOperationException("CRC mismatch in message");
            }

            var payload = new byte[length];
            payloadSpan.CopyTo(payload);

            onMessage(MessageWrapper.From((MessageType)type, payload));
            
            _stream.Position += frameSize;
        }

        // Compact remaining bytes
        if (_stream.Position > 0)
        {
            var remaining = _stream.Length - _stream.Position;

            if (remaining > 0)
            {
                Array.Copy(
                    _stream.GetBuffer(),
                    _stream.Position,
                    _stream.GetBuffer(),
                    0,
                    remaining);
            }

            _stream.SetLength(remaining);
            _stream.Position = 0;
        }
    }
}